Topics/Development_Process/GraphQL_VS_REST.md

```diff
@@ -0,0 +1,94 @@
+
+
+## GraphQL Overview
+GraphQL is difference from REST in the sense that it is not actually an API architecture. In fact, it is actually a `query language`, hence the QL in its name. This leads to a number of distinct characteristics that we will discuss:
+1. In GraphQL there are only two types of requests: `Queries`, and `Mutations`. A query is used to get data, i.e, "query" some data. Everything else would be a mutation; creating, updating, or deleting data all mutate, and thus they are mutations.
+    + To follow proper GraphQL standards, any endpoint that only sends back data would fall under `query`. Any endpoint that changes data in any way would fall under `mutation`.
+2. GraphQL is very specific; when you call an endpoint, you can request *exactly* what data you want, and you can send different structures of data, based on customizable `types`.
+3. GraphQL only requires one endpoint. Usually, this is found at the `/graphql` URL on a webapp. Since GraphQL is a query language, whenever call the endpoint, you use the request body to specify your query. From this, you can define mutations and queries, and can in fact, get multiple pieces on data through one request. Using the same example from REST, you would only need to send one `/graphql` request, and just write both queries for data in the same request. This allows you to get the same amount of data using one request instead of two separate ones.
+
+
+## When does REST make sense to use?
+REST makes the most sense if you deal with simple data, and you do not need to only get specific pieces of data; getting all of what you need is okay. In other words, if you're getting a User's data, for instance, you are okay with getting all of their data, instead of just their username for instance.
+
+## When does GraphQL make sense to use?
+GraphQL makes sense if you are dealing with quite complicated data, especially if it is large swaths of data where you might only need a specific chunk of it. It also makes sense if you need to get numerous different pieces of data at the same time, and want to do so efficiently. GraphQL is much faster than REST in terms of actual efficiency of getting data, but has a trade-off of a complex set up compared to REST.
+
+
+## Quick reference for common uses
+In JavaScript, you can implement REST APIs through ExpressJS. Here is an example request, where you would call an already existing API request using fetch(). 
+```javascript
+const response = await fetch("http://example.com/movies.json", {
+  method: "POST",
+  body: JSON.stringify({
+    name: "Barbie",
+    director: "Greta Gerwig",
+    stars: [
+        "Margot Robbie"
+    ]
+  })
+});
+```
+
+You can implement GraphQL in JavaScript with the GraphQL library. Here is an example GraphQL query, where you would get the director of the Barbie movie, similarly to the above request.
+```graphql
+query {
+  getMovie(params: {
+    name: "Barbie"
+  }) 
+  { director }
+}
+```

```

Comment:
I think this concrete example of comparing REST API and GraphQL is really great and definitely makes things very clear!

Reply:
Glad to hear :)